src/twisted/internet/cfreactor.py

```diff
@@ -44,7 +44,8 @@
 _preserveSOError = 1 << 6
 
 
-class _WakerPlus(_Waker):
+
+class _WakerPlus(_Waker):  # type: ignore[valid-type,misc]
```

Comment:
I don't understand why the extra typing comment is required. _Waker should now by any _IWaker and _WakerPlus implements that interface.

Reply:
Without that comment, mypy will give this error in this branch:

```
src/twisted/internet/cfreactor.py:48:18: error: Variable "twisted.internet.posixbase._Waker" is not valid as a type  [valid-type]
src/twisted/internet/cfreactor.py:48:18: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
src/twisted/internet/cfreactor.py:48:18: error: Invalid base class "_Waker"  [misc]
```

`_Waker` is not a concrete type according to the static checker, because here: https://github.com/twisted/twisted/pull/1329/files#diff-efe69815061dceb870b109751c0bd433L192
_Waker can be a `_UnixWaker` or a `_SocketWaker`.  Even though with this patch both those classes `@implementer _IWaker`, from the static checker's perspective, those are two different classes.  So `_WakerPlus` either inherits from one of those classes or another, depending on the platform.  So `_Waker` is a dynamic base class, which mypy does not support.

Reply:
what a mess... not sure if this is a mypy or mypy-zope bug... or I am not using mypy-zope the right way

to me, it looks like a mypy-zope bug

Reading the linked docs, it looks like _Waker need an explicit type definition.

My understanding is that the following should do the job and is the-right-way(tm) to have the type defined

```diff
diff --git a/src/twisted/internet/posixbase.py b/src/twisted/internet/posixbase.py
index 9bda9c9a4..49390637f 100644
--- a/src/twisted/internet/posixbase.py
+++ b/src/twisted/internet/posixbase.py
@@ -223,8 +223,9 @@ class _UnixWaker(_FDWaker):
 
 
 
+_Waker: _IWaker
 if platformType == 'posix':
-    _Waker = _UnixWaker  # type: Type[_IWaker]
+    _Waker = _UnixWaker
 else:
     # Primarily Windows and Jython.
     _Waker = _SocketWaker
```

Reply:
I have created this report https://github.com/Shoobx/mypy-zope/issues/19

Reply:
ok... so it looks like mypy is ignored  when we have a variable type alias :)  ... just variable types are supported... and just alias types are supported... but when you combine them things get messy.

My take is to consider _Waker a type alias and to ignore the variable assignment error.


the diff would be as below.

In this way _Waker is a valid type

```diff
diff --git a/src/twisted/internet/cfreactor.py b/src/twisted/internet/cfreactor.py
index 4c9d0b60a..2de8ccb22 100644
--- a/src/twisted/internet/cfreactor.py
+++ b/src/twisted/internet/cfreactor.py
@@ -45,7 +45,7 @@ _preserveSOError = 1 << 6
 
 
 
-class _WakerPlus(_Waker):  # type: ignore[valid-type,misc]
+class _WakerPlus(_Waker):
     """
     The normal Twisted waker will simply wake up the main loop, which causes an
     iteration to run, which in turn causes L{ReactorBase.runUntilCurrent}
diff --git a/src/twisted/internet/posixbase.py b/src/twisted/internet/posixbase.py
index 9bda9c9a4..fc1c83a64 100644
--- a/src/twisted/internet/posixbase.py
+++ b/src/twisted/internet/posixbase.py
@@ -224,10 +224,10 @@ class _UnixWaker(_FDWaker):
 
 
 if platformType == 'posix':
-    _Waker = _UnixWaker  # type: Type[_IWaker]
+    _Waker = _UnixWaker
 else:
     # Primarily Windows and Jython.
-    _Waker = _SocketWaker
+    _Waker = _SocketWaker  # type: ignore[misc,assignment]
 
 
```

Reply:
OK, I incorporated this patch.

Reply:
with this path, all _Waker will be _UnixWaker ... not really what we want... but I have no idea how to define it in mypy